test/utils/test_normalization.py

```diff
@@ -0,0 +1,168 @@
+        [1, 0, 0, 0, 1, 0],
+    ]
+
+    normalized_A_opt = compute_kipf_adjacency_normalized_matrix(csr_matrix(A_opt))
+    expected_result = [
+        [0.0, 0.5, 0.0, 0.0, 0.0, 0.5],
+        [0.5, 0.0, 0.5, 0.0, 0.0, 0.0],
+        [0.0, 0.5, 0.0, 0.5, 0.0, 0.0],
+        [0.0, 0.0, 0.5, 0.0, 0.5, 0.0],
+        [0.0, 0.0, 0.0, 0.5, 0.0, 0.5],
+        [0.5, 0.0, 0.0, 0.0, 0.5, 0.0],
+    ]
+
+    assert np.allclose(normalized_A_opt.toarray(), expected_result)
+
+    normalized_A_opt = compute_kipf_adjacency_normalized_matrix(
+        csr_matrix(A_opt), add_identity=True
+    )
+    expected_result = [
+        [0.33333333, 0.33333333, 0.0, 0.0, 0.0, 0.33333333],
+        [0.33333333, 0.33333333, 0.33333333, 0.0, 0.0, 0.0],
+        [0.0, 0.33333333, 0.33333333, 0.33333333, 0.0, 0.0],
+        [0.0, 0.0, 0.33333333, 0.33333333, 0.33333333, 0.0],
+        [0.0, 0.0, 0.0, 0.33333333, 0.33333333, 0.33333333],
+        [0.33333333, 0.0, 0.0, 0.0, 0.33333333, 0.33333333],
+    ]
+
+    assert np.allclose(normalized_A_opt.toarray(), expected_result)
+
+
+def test_compute_bunch_normalized_matrices():
+    """Unit tests for bunch_normalization function."""
+    # Test case 1: Normalization with numpy arrays
+    B1 = np.array([[1, 0, 1], [0, 1, 1], [1, 1, 0]])
+    B2 = np.array([[1, 0, 1], [0, 1, 0], [1, 1, 1]])
+    B1N, B1TN, B2N, B2TN = compute_bunch_normalized_matrices(B1, B2)
+
+    assert np.allclose(
+        B1N,
+        np.array([[0.1, 0.0, 0.1], [0.0, 0.125, 0.125], [0.16666667, 0.16666667, 0.0]]),
+    )
+
+    assert np.allclose(
+        B1TN,
+        np.array([[0.2, 0.0, 0.33333333], [0.0, 0.125, 0.16666667], [0.3, 0.375, 0.0]]),
+    )
+
+    assert np.allclose(
+        B2N,
+        np.array(
+            [
+                [0.33333333, 0.0, 0.33333333],
+                [0.0, 0.33333333, 0.0],
+                [0.33333333, 0.33333333, 0.33333333],
+            ]
+        ),
+    )
+
+    assert np.allclose(
+        B2TN,
+        np.array(
+            [[0.5, 0.0, 0.33333333], [0.0, 1.0, 0.33333333], [0.5, 0.0, 0.33333333]]
+        ),
+    )
+
+    # Test case 2: Normalization with scipy coo_matrices
+    B1 = csr_matrix([[1, 0, 1], [0, 1, 1], [1, 1, 0]])
+    B2 = csr_matrix([[1, 0, 1], [0, 1, 0], [1, 1, 1]])
+    B1N, B1TN, B2N, B2TN = compute_bunch_normalized_matrices(B1, B2)
+
+    assert np.allclose(
+        B1N.toarray(),
+        [[0.1, 0.0, 0.1], [0.0, 0.125, 0.125], [0.16666667, 0.16666667, 0.0]],
+    )
+
+    assert np.allclose(
+        B1TN.toarray(),
+        [[0.2, 0.0, 0.33333333], [0.0, 0.125, 0.16666667], [0.3, 0.375, 0.0]],
+    )
+
+    assert np.allclose(
+        B2N.toarray(),
+        [
+            [0.33333333, 0.0, 0.33333333],
+            [0.0, 0.33333333, 0.0],
+            [0.33333333, 0.33333333, 0.33333333],
+        ],
+    )
+
+    assert np.allclose(
+        B2TN.toarray(),
+        np.array(
+            [[0.5, 0.0, 0.33333333], [0.0, 1.0, 0.33333333], [0.5, 0.0, 0.33333333]]
+        ),
```

Comment:
There are a lot of distinct test cases that have been put together under one test function. Please separate them. Best practice is to have distinct test cases for different tests for easy isolation of issues, and debugging